metagraph/src/common/vectors/sd_vector_disk.hpp

```diff
@@ -0,0 +1,530 @@
+#ifndef __SD_VECTOR_DISK__
+#define __SD_VECTOR_DISK__
+
+#include <type_traits>
+
+#include <sdsl/sd_vector.hpp>
+#include <sdsl/int_vector.hpp>
+#include <sdsl/int_vector_buffer.hpp>
+#include <sdsl/select_support_mcl.hpp>
+#include <sdsl/util.hpp>
+#include <sdsl/iterators.hpp>
+
+// Based on https://github.com/simongog/sdsl-lite/blob/master/include/sdsl/sd_vector.hpp
+
+// FYI: This implementation does NOT support multithreading
+
+namespace sdsl {
+
+// forward declaration needed for friend declaration
+template<uint8_t t_b          = 1,
+         class t_hi_bit_vector= bit_vector,
+         class t_select_1     = typename t_hi_bit_vector::select_1_type,
+         class t_select_0     = typename t_hi_bit_vector::select_0_type>
+class rank_support_sd_disk;  // in sd_vector
+
+// forward declaration needed for friend declaration
+template<uint8_t t_b          = 1,
+         class t_hi_bit_vector= bit_vector,
+         class t_select_1     = typename t_hi_bit_vector::select_1_type,
+         class t_select_0     = typename t_hi_bit_vector::select_0_type>
+class select_support_sd_disk;  // in sd_vector
+
+//! Class for in-place construction of sd_vector from a strictly increasing sequence
+/*! \par Building an sd_vector will clear the builder.
+ */
+class sd_vector_disk_builder
+{
+        template<typename, typename, typename>
+        friend class sd_vector_disk;
+
+    public:
+        typedef bit_vector::size_type size_type;
+
+    private:
+        size_type m_size, m_capacity;
+        uint8_t   m_wl;
+        size_type m_tail, m_items;
+        size_type m_last_high, m_highpos;
+
+        std::string filename;
+        size_t file_offset;
+        int_vector_buffer<> m_low;
+        bit_vector   m_high;
+
+    public:
+        sd_vector_disk_builder() :
+            m_size(0), m_capacity(0),
+            m_wl(0),
+            m_tail(0), m_items(0),
+            m_last_high(0), m_highpos(0) {}
+
+        //! Constructor
+        /*! \param n Vector size.
+         *  \param m The number of 1-bits.
+         *  \param filename The buffer filename to where the bitmap will eventually be stored.
+         */
+        sd_vector_disk_builder(size_type n, size_type m, const std::string &fname, size_t offset = 0) :
+            m_size(n), m_capacity(m),
+            m_wl(0),
+            m_tail(0), m_items(0),
+            m_last_high(0), m_highpos(0), filename(fname), file_offset(offset)
+        {
+            if(m_capacity > m_size)
+                throw std::runtime_error("sd_vector_builder: requested capacity is larger than vector size.");
+
+            size_type logm = bits::hi(m_capacity) + 1;
+            const size_type logn = bits::hi(m_size) + 1;
+            if(logm == logn)
+            {
+                --logm; // to ensure logn-logm > 0
+                assert(logn - logm > 0);
+            }
+            m_wl = logn - logm;
```

Comment:
does this class still work if compressing a vector where all bits are set?

Reply:
Yeah, it should. And it's in the tests, isn't it?

Reply:
I couldn't find a test like that in sdsl. The tests I found fill the vectors randomly, e.g.,
https://github.com/simongog/sdsl-lite/blob/master/test/bit_vector_test.cpp

We do have this in metagraph:
https://github.com/ratschlab/metagraph/blob/master/metagraph/tests/test_bit_vector.cpp#L195

But the problem is that it's tested with `bit_vector_sd`, which I believe defaults to storing the bit-flipped version of the vector if the density is high enough. So again, I'm not sure if this gets tested.

Reply:
There are specific tests for sd_vector_disk. BTW, it's implemented in metagraph, not sdsl. Hence tested here.
Also, note that it's never inverted.
https://github.com/ratschlab/metagraph/pull/433/files#diff-44088ac3b4ca1eb7a80c3d8cdc12044299432149e4d3eb23b6c6cda0aea22ba6R637

Reply:
ok, looks good

Reply:
thanks for the clarification